parsec/backend/postgresql/pki.py

```diff
@@ -0,0 +1,439 @@
+        AND enrollment_id=$enrollment_id
+    )
+    """
+)
+
+
+_q_get_user_from_device_id = Q(
+    f"""
+    SELECT *
+    FROM user_
+    WHERE
+        user_.organization = { q_organization_internal_id("$organization_id") }
+        AND user_._id=(
+            SELECT user_
+            FROM device
+            WHERE device._id=$device_id
+            AND device.organization = { q_organization_internal_id("$organization_id") }
+        )
+    LIMIT 1
+    """
+)
+
+
+def _build_enrollment_info(entry) -> PkiEnrollmentInfo:
+    if entry["enrollment_state"] == PkiEnrollmentStatus.SUBMITTED.value:
+        return PkiEnrollmentInfoSubmitted(
+            enrollment_id=entry["enrollment_id"], submitted_on=entry["submitted_on"]
+        )
+    elif entry["enrollment_state"] == PkiEnrollmentStatus.CANCELLED.value:
+        return PkiEnrollmentInfoCancelled(
+            enrollment_id=entry["enrollment_id"],
+            submitted_on=entry["submitted_on"],
+            cancelled_on=entry["info_cancelled"]["cancelled_on"],
+        )
+    elif entry["enrollment_state"] == PkiEnrollmentStatus.REJECTED.value:
+        return PkiEnrollmentInfoRejected(
+            enrollment_id=entry["enrollment_id"],
+            submitted_on=entry["submitted_on"],
+            rejected_on=entry["info_rejected"]["rejected_on"],
+        )
+    elif entry["enrollment_state"] == PkiEnrollmentStatus.ACCEPTED.value:
+        return PkiEnrollmentInfoAccepted(
+            enrollment_id=entry["enrollment_id"],
+            submitted_on=entry["submitted_on"],
+            accepted_on=entry["info_accepted"]["accepted_on"],
+            accepter_der_x509_certificate=entry["info_accepted"]["accepter_der_x509_certificate"],
+            accept_payload_signature=entry["info_accepted"]["accept_payload_signature"],
+            accept_payload=entry["info_accepted"]["accept_payload"],
+        )
+    else:
+        assert False
+
+
+class PGPkiEnrollmentComponent(BasePkiEnrollmentComponent):
+    def __init__(self, dbh: PGHandler):
+        self.dbh = dbh
+
+    def register_components(self, **other_components):
+        self._user_component = other_components["user"]
+
+    async def submit(
+        self,
+        organization_id: OrganizationID,
+        enrollment_id: UUID,
+        force: bool,
+        submitter_der_x509_certificate: bytes,
+        submit_payload_signature: bytes,
+        submit_payload: bytes,
+        submitted_on: DateTime,
+    ) -> None:
+        """
+        Raises:
+            PkiEnrollmentCertificateAlreadySubmittedError
+            PkiEnrollmentAlreadyEnrolledError
+        """
+        submitter_der_x509_certificate_sha1 = hashlib.sha1(submitter_der_x509_certificate).digest()
+        async with self.dbh.pool.acquire() as conn, conn.transaction():
+            # Assert enrollment_id not used
+            row = await conn.fetchrow(
+                *_q_get_pki_enrollment_for_update(
```

Comment:
FOR UPDATED is not mandatory here (given we raise just after we get a result, and if we don't get a result there is nothing to lock so we are not protected against concurrent insertion, see my following comment about `UniqueViolationError`)

However there is nothing wrong with adding the FOR UPDATE, so I'm okay with using this here ;-)
(but maybe add a small comment)